File: gfapy/alignment/cigar.py
import re
import gfapy

class CIGAR(list):
  """
  Representation of the contents of a CIGAR string.

  Each operation is represented by a
  :class:`CIGAR.Operation <gfapy.alignment.cigar.CIGAR.Operation>`,
  which specifies an operation length and operation symbol.

  Instances are usually created from their string representations, using the
  :class:`~gfapy.alignment.alignment.Alignment` factory class constructor.

  Warning:
    Although the GFA1 specification does not forbid the
    operation symbols NSHX=, these are not allowed in GFA2
    and thus their use in GFA1 is discouraged.
  """

  def complement(self):
    """The CIGAR when switching the role of the two aligned segments.

    Example:
      >>> import gfapy
      >>> str(gfapy.Alignment("2M1D3M").complement())
      '3M1I2M'

    Returns:
      CIGAR: the complement CIGAR
    """
    comp = list(reversed(self))
    for op in comp:
      if   op.code == "I": op.code = "D"
      elif op.code == "S": op.code = "D"
      elif op.code == "D": op.code = "I"
      elif op.code == "N": op.code = "I"
    return CIGAR(comp)

  def validate(self, version = "gfa1"):
    """Validates the instance.

    Parameters:
      version (str): 'gfa1' or 'gfa2'

    Raises:
      ~gfapy.error.VersionError: If a wrong **version** is specified.
      ~gfapy.error.TypeError: If a component of the list is not a
          CIGAR Operation; If the CIGAR operation length is not an integer or
          a string representing an integer.
      ~gfapy.error.ValueError: If the length of an operation is < 0; If an
          operation code is invalid in general or for the specified GFA version.
    """
    if version != "gfa1" and version != "gfa2":
      raise gfapy.VersionError(
          "Version error: {}".format(repr(version)))
    for op in self:
      if not isinstance(op, gfapy.CIGAR.Operation):
        raise gfapy.TypeError(
            "Element is not a CIGAR operation: {}\n".format(op)+
            "CIGAR instance is invalid: {}".format(self))
      op.validate(version = version)

  def length_on_reference(self):
    """Length of the aligned substring on the reference sequence
    (**from** sequence for GFA1 links/containments;
    **sid1** sequence for GFA2 edges)

    Returns:
      int
    """
    l = 0
    for op in self:
      if op.code in ["M", "=", "X", "D" , "N"]:
        l += op.length
    return l

  def length_on_query(self):
    """
    Lenght of the aligned substring on the query sequence
    (**to** sequence for GFA1 links/containments;
    **sid2** sequence for GFA2 edges)

    Returns:
      int
    """
    l = 0
    for op in self:
      if op.code in ["M", "=", "X", "I", "S"]:
        l += op.length
    return l

  @classmethod
  def _from_string(cls, string, valid = False, version = "gfa1"):
    """Create a CIGAR instance from its string representation.

    Parameters:
      string (str)
      valid (bool): If **True** the string is guaranteed to be valid.
        (Defaults to **False**)
      version (str): 'gfa1' or 'gfa2'

    Returns:
      ~gfapy.alignment.cigar.CIGAR or
      ~gfapy.alignment.placeholder.AlignmentPlaceholder

    Raises:
      ~gfapy.error.FormatError: If the string is not a valid CIGAR string.
    """
    if string == "*":
      return gfapy.AlignmentPlaceholder()
    cigar = CIGAR()
    if not valid:
      if version == "gfa1":
        if not re.match(r"^([0-9]+[MIDNSHPX=])+$", string):
          raise gfapy.FormatError()
      elif version == "gfa2":
        if not re.match(r"^([0-9]+[MIDP])+$", string):
          raise gfapy.FormatError()
    for m in re.finditer("([0-9]+)([MIDNSHPX=])", string):
      cigar.append(CIGAR.Operation(int(m.group(1)), m.group(2)))
    return cigar

  def __str__(self):
    if not self:
      return "*"
    else:
      return "".join([str(op) for op in self])

  def __repr__(self):
    return "gfapy.CIGAR([{}])".format(", ".join([repr(op) for op in self]))

  class Operation:
    """An operation in a CIGAR string.

    Attributes:
      ~Operation.length (int): Operation length.
      code (str): Operation code, one of
          :attr:`~Operation.CODE`.
    """

    CODE_GFA1_ONLY = ["S", "H", "N", "X", "="]
    """Operations only valid in GFA1"""

    CODE_GFA1_GFA2 = ["M", "I", "D", "P"]
    """Operations valid in GFA1 and GFA2"""

    CODE = CODE_GFA1_ONLY + CODE_GFA1_GFA2
    """CIGAR operation codes"""

    def validate(self, version = "gfa1"):
      """Validates the CIGAR operation.

      Parameters:
        version (str): 'gfa1' or 'gfa2'

      Raises:
        ~gfapy.error.VersionError: If a wrong **version** is specified.
        ~gfapy.error.TypeError: If the CIGAR operation length is not an integer
            or a string representing an integer.
        ~gfapy.error.ValueError: If the length of an operation is < 0; If an
            operation code is invalid in general or for the specified GFA
            version.
      """
      if version != "gfa1" and version != "gfa2":
        raise gfapy.VersionError(
            "Version error: {}".format(repr(version)))
      if not isinstance(self.length, int) and not isinstance(self.length, str):
        raise gfapy.TypeError(
            "Type error: length of CIGAR is {}".format(self.length))
      if(int(self.length) < 0):
        raise gfapy.ValueError("Length of CIGAR is {}".format(self.length))
      if version == "gfa2":
        if not self.code in Operation.CODE_GFA1_GFA2:
          raise gfapy.ValueError()
      else:
        if not self.code in Operation.CODE:
          raise gfapy.ValueError()

    def __init__(self, length, code):
      self.length = length
      self.code = code

    def __len__(self):
      return self.length

    def __str__(self):
      return "{}{}".format(self.length, self.code)

    def __repr__(self):
      return "gfapy.CIGAR.Operation({},{})".format(self.length, repr(self.code))

    def __eq__(self, other):
      return self.length == other.length and self.code == other.code

Operation = CIGAR.Operation

File: lib/rgfa/alignment/cigar.rb
RGFA::Alignment ||= Module.new

# Array of {RGFA::Alignment::CIGAR::Operation CIGAR operations}.
# Represents the contents of a CIGAR string.
# @tested_in api_alignment
class RGFA::Alignment::CIGAR < Array

  # Compute the CIGAR for the segments when these are switched.
  #
  # @example Computing the complement CIGAR
  #
  #   "2M1D3M".to_alignment.complement.to_s
  #   # => "3M1I2M"
  #
  #   # S1 + S2 + 2M1D3M
  #   #
  #   # S1+  ACGACTGTGA
  #   # S2+      CT-TGACGG
  #   #
  #   # S2-  CCGTCA-AG
  #   # S1-     TCACAGTCGT
  #   #
  #   # S2 - S1 - 3M1I2M
  #
  # @return [RGFA::Alignment::CIGAR]
  def complement
    RGFA::Alignment::CIGAR.new(clone.reverse.map do |op|
      if op.code == :I or op.code == :S
        op.code = :D
      elsif op.code == :D or op.code == :N
        op.code = :I
      end
      op
    end)
  end

  # String representation of the CIGAR
  # @note no validation is performed, use #validate if required
  # @return [String] CIGAR string
  def to_s
    placeholder? ? "*" : (map(&:to_s).join)
  end

  # Validate the instance
  # @param version [Symbol] <i>(defaults to: +:gfa1+)</i> if :gfa2,
  #   then only CIGAR codes M/I/D/P are allowed, if :gfa1 all CIGAR codes
  # @raise [RGFA::ValueError] if a code is invalid or a length is negative
  # @raise [RGFA::TypeError] if a length is not an Integer or
  #   the array contains anything which is not interpretable as a
  #   cigar operation
  # @raise [RGFA::VersionError] if a wrong version is provided
  # @return [void]
  def validate(version: :gfa1)
    if ![:gfa1, :gfa2].include?(version)
      raise RGFA::VersionError, "Version unknown: #{version}"
    end
    any? do |op|
      begin
        op = op.to_cigar_operation
      rescue
        raise RGFA::TypeError, "Array contains elements which are "+
          "not CIGAR operations: #{self.inspect}"
      end
      op.validate(version: version)
    end
  end

  # @param valid [nil] ignored, for compatibility
  # @param version [nil] ignored, for compatibility
  # @return [RGFA::Alignment::CIGAR] self
  def to_alignment(valid: nil, version: :nil)
    self
  end

  # Lenght of the aligned substring on the reference sequence
  # (+from+ sequence for GFA1 links/containments;
  #  +sid1+ sequence for GFA2 edges)
  # @return [Integer] length of the aligned substring on the
  #   reference sequence
  def length_on_reference
    l = 0
    each do |op|
     if [:M, :"=", :X, :D, :N].include?(op.code)
       l += op.len
     end
    end
    return l
  end

  # Lenght of the aligned substring on the query sequence
  # (+to+ sequence for GFA1 links/containments;
  #  +sid2+ sequence for GFA2 edges)
  # @return [Integer] length of the aligned substring on the
  #   query sequence
  def length_on_query
    l = 0
    each do |op|
     if [:M, :"=", :X, :I, :S].include?(op.code)
       l += op.len
     end
    end
    return l
  end

  # Create a deep copy
  # @return [RGFA::Alignment::CIGAR]
  def clone
    RGFA::Alignment::CIGAR.new(map{|x|x.clone})
  end

  # @api private
  # @tested_in unit_alignment
  module API_PRIVATE

    # @return [RGFA::Alignment::CIGAR] self
    # @param valid [nil] ignored, for compatibility
    # @param version [nil] ignored, for compatibility
    def to_cigar(valid: nil, version: :nil)
      self
    end

    module ClassMethods

      # Parse a CIGAR string into an array of CIGAR operations.
      #
      # Each operation is represented by a {RGFA::Alignment::CIGAR::Operation},
      # i.e. a tuple of operation length and operation code.
      #
      # The operation code is one of MIDP for GFA2 or MIDPNSHX= for GFA1.
      # The additional operations allowed in GFA1 have an unclear meaning
      # in the context of GFA and should be avoided.
      #
      # @param version [Symbol] <i>(defaults to: +gfa1+)</i> if +:gfa2+,
      #   then only CIGAR codes M/I/D/P are allowed, if +:gfa1+ all CIGAR codes
      # @param valid [Boolean] <i>(defaults to: +false+)</i> if +true+,
      #   the string is guaranteed to be valid
      # @raise [RGFA::FormatError] if the string is not a valid CIGAR string
      # @raise [RGFA::VersionError] if a wrong version is provided
      # @return [RGFA::Alignment::CIGAR]
      def from_string(str, valid: false, version: :gfa1)
        a = RGFA::Alignment::CIGAR.new
        unless valid
          case version
          when :gfa1
            if str !~ /^([0-9]+[MIDPNSHX=])+$/
              raise RGFA::FormatError,
              "The string #{str} does not represent a valid CIGAR string"
            end
          when :gfa2
            if str !~ /^([0-9]+[MIDP])+$/
              raise RGFA::FormatError,
              "The string #{str} does not represent a valid GFA2 CIGAR string"
            end
          else
            raise RGFA::VersionError, "Version unknown: #{version}"
          end
        end
        str.scan(/[0-9]+[MIDPNSHX=]/).each do |op|
          len = op[0..-2].to_i
          code = op[-1..-1].to_sym
          a << RGFA::Alignment::CIGAR::Operation.new(len, code)
        end
        return a
      end

    end

  end
  include API_PRIVATE
  extend API_PRIVATE::ClassMethods

end

# An operation in a CIGAR string
# @tested_in api_alignment
class RGFA::Alignment::CIGAR::Operation

  # @!attribute [rw] len
  #   @return [Integer > 0] operation length
  attr_accessor :len

  # @!attribute [rw] code
  #   @return [RGFA::Alignment::CIGAR::Operation::CODE] operation code
  attr_accessor :code

  # CIGAR operation code
  CODE_GFA1_ONLY = [:S, :H, :N, :X, :"="]
  CODE_GFA1_GFA2 = [:M, :I, :D, :P]
  CODE = CODE_GFA1_ONLY + CODE_GFA1_GFA2

  # @param len [Integer] length of the operation
  # @param code [RGFA::Alignment::CIGAR::Operation::CODE] code of the operation
  def initialize(len, code)
    @len = len
    @code = code
  end

  # The string representation of the operation
  # @note no validation is performed, use #validate if required
  # @return [String]
  def to_s
    "#{len}#{code}"
  end

  # Compare two operations
  # @return [Boolean]
  def ==(other)
    other.len == len and other.code == code
  end

  # Validate the operation
  # @param version [Symbol] <i>(defaults to: +:gfa1+)</i> if :gfa2,
  #   then only CIGAR codes M/I/D/P are allowed, if :gfa1 all CIGAR codes
  # @raise [RGFA::ValueError] if the code is invalid or the length is negative
  # @raise [RGFA::TypeError] if the length is not an Integer
  # @raise [RGFA::VersionError] if a wrong version is provided
  # @return [void]
  def validate(version: :gfa1)
    if ![:gfa1, :gfa2].include?(version)
      raise RGFA::VersionError, "Version unknown: #{version}"
    end
    begin
      len = Integer(@len)
    rescue
      raise RGFA::TypeError, "CIGAR operation: #{self.inspect}\n"+
        "CIGAR length cannot be casted to Integer (class: #{len.class})"
    end
    begin
      code = @code.to_sym
    rescue
      raise RGFA::TypeError, "CIGAR operation: #{self.inspect}\n"+
        "CIGAR code cannot be casted to symbol (class: #{code.class})"
    end
    if len < 0
      raise RGFA::ValueError,
        "Length of CIGAR operation #{self} is invalid (#{len})"
    elsif RGFA::Alignment::CIGAR::Operation::CODE_GFA1_ONLY.include?(code)
      if version == :gfa2
        raise RGFA::ValueError, "CIGAR operation: #{self.inspect}\n"+
          "CIGAR code is not supported in GFA2: #{code}"
      end
    elsif !RGFA::Alignment::CIGAR::Operation::CODE_GFA1_GFA2.include?(code)
      raise RGFA::ValueError, "CIGAR operation: #{self.inspect}\n"+
        "Invalid CIGAR code found: #{code}"
    end
  end

  # @api private
  # @tested_in unit_alignment
  module API_PRIVATE
    # @return [RGFA::Alignment::CIGAR::Operation] self
    def to_cigar_operation
      self
    end
  end
  include API_PRIVATE

end

class Array

  # @api private
  # @tested_in unit_alignment
  module API_PRIVATE
    # Create a {RGFA::Alignment::CIGAR} instance from the content of the array.
    # @param valid [nil] ignored, for compatibility
    # @param version [nil] ignored, for compatibility
    # @return [RGFA::Alignment::CIGAR]
    def to_cigar(valid: nil, version: nil)
      RGFA::Alignment::CIGAR.new(self)
    end

    # Create a {RGFA::Alignment::CIGAR::Operation} instance
    # from the array content
    # @return [RGFA::Alignment::CIGAR::Operation]
    def to_cigar_operation
      RGFA::Alignment::CIGAR::Operation.new(Integer(self[0]), self[1].to_sym)
    end
  end
  include API_PRIVATE

end

class String

  # @api private
  # @tested_in unit_alignment
  module API_PRIVATE
    # Parse CIGAR string
    # @return [RGFA::Alignment::CIGAR,RGFA::Alignment::Placeholder]
    #    CIGAR or Placeholder (if +*+)
    # @param valid [Boolean] <i>(defaults to: +false+)</i> if +true+,
    #   the string is guaranteed to be valid
    # @param version [Symbol] <i>(defaults to: +:gfa1+)</i> if :gfa2,
    #   then only CIGAR codes M/I/D/P are allowed, if :gfa1 all CIGAR codes
    # @raise [RGFA::ValueError] if the string is not a valid CIGAR string
    # @raise [RGFA::VersionError] if a wrong version is provided
    def to_cigar(valid: false, version: :gfa1)
      if placeholder?
        return RGFA::Alignment::Placeholder.new
      else
        return RGFA::Alignment::CIGAR.from_string(self, valid: valid,
                                                  version: version)
      end
    end
  end
  include API_PRIVATE

end
